File: CSES/Sorting_and_Searching/1164_Room_Allocation.py
import sys
input = sys.stdin.readline
def sol(arr, n):
    arr2 = []
    for i in range(n):
        arr2.append((arr[i][0], 0, i))
        arr2.append((arr[i][1], 1, i))
    arr2.sort()
    cnt = 0
    empt = []
    res = [0] * n
    for i, j, k in arr2:
        if j == 0:
            if not empt:
                cnt += 1
                res[k] = cnt
            else:
                res[k] = empt.pop()
        else:
            empt.append(res[k])
    
    print(cnt)
    for i in range(n):
        print(res[i], end=' ')
    print()
        
n = int(input())
arr = []
for i in range(n):
    arr.append(list(map(int,input().split())))
(sol(arr, n))

File: CSES/Sorting_and_Searching/1164_Room_Allocation.cpp
#include <iostream>
#include <vector>
#include <algorithm>

#define pii pair<int, int>
#define PB push_back
#define MP make_pair

using namespace std;

int main() {
    cout.sync_with_stdio(false);
    cin.tie(nullptr);
    int n, k, i, j, cnt = 0;
    string s;
    cin >> n;
    vector<pii> arr;
    vector<int> res(n, 0), empt;
    for (i = 0; i < n; ++i) {
        cin >> j >> k;
        arr.PB(MP(j, i));
        arr.PB(MP(k, i + n));
    }
    sort(arr.begin(), arr.end());
    for (auto &[j, i]: arr) {
        if (i < n && res[i] == 0)
            if (empt.empty()) res[i] = ++cnt;
            else res[i] = empt.back(), empt.pop_back();
        else empt.PB(res[i - n]);
    }
    cout << cnt << '\n';
    for (auto &i: res)
        cout << i << ' ';

    return 0;
}
